In 401/bulk table actions

SaraBCL opened the pull request:
- Adds an assign button to assign a selection of categories to selected inputs
- Adds an unassign button to remove selected inputs from the current category 
- Moves the title of the table to separate component (felt it was needed for clarity, index was growing too much otherwise)

guilherme-andrade committed:
[CL2-6263] changed color to clBlue

Commit:
idea edit form wrong locale bugfix

kogre committed:
New translations en.json (Romanian)

kogre committed:
New translations en.yml (Danish)

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (Dutch, Belgium)

kogre committed:
New translations en.json (Serbian (Latin))

kogre committed:
New translations en.json (English, United Kingdom)

kogre committed:
New translations en.json (English, Canada)

kogre committed:
New translations en.json (Maori)

kogre committed:
New translations en.json (Spanish, Chile)

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.json (Swedish)

kogre committed:
New translations en.json (Greenlandic)

kogre committed:
New translations en.json (Polish)

kogre committed:
New translations en.json (Norwegian)

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Hungarian)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (Danish)

kogre committed:
New translations en.json (Arabic)

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (French, Belgium)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (English)

kogre committed:
Update source file en.json

guilherme-andrade committed:
Update README.md

SaraBCL committed:
updates changelog

SaraBCL committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into IN-401/bulk-table-actions

SaraBCL committed:
adds select mechanism

SaraBCL committed:
moves buttons from user table to the component library

SaraBCL committed:
adds non-functional buttons

SaraBCL committed:
Merge branch 'IN-403/filter-by-category' of github.com:CitizenLabDotCo/citizenlab into IN-401/bulk-table-actions

SaraBCL committed:
Moves Action Buttons next to the title

SaraBCL committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into IN-401/bulk-table-actions

SaraBCL committed:
Makes assignment functional

SaraBCL committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into IN-401/bulk-table-actions

SaraBCL committed:
Makes removal fuctional

SaraBCL committed:
Don't show category selection when there is no category to add.

SaraBCL committed:
Minor style and type fixes

Commit:
Translations updated by CI (extract-intl)

SaraBCL committed:
Resets selection on page change

SaraBCL committed:
Merge branch 'IN-401/bulk-table-actions' of github.com:CitizenLabDotCo/citizenlab into IN-401/bulk-table-actions

SaraBCL committed:
Fixes existing tests

SaraBCL committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into IN-401/bulk-table-actions

SaraBCL committed:
Updates component library

SaraBCL committed:
Adds tests for action buttons

IvaKop commented on front/app/modules/commercial/insights/admin/containers/Insights/Edit/InputsTable/index.tsx:
```diff
@@ -105,20 +135,45 @@ const InputsTable = ({
       ) : (
         <>
+          <TitleRow>
+            <TableTitle />
```
We have a problem here with the empty state and the table title. Even when there are no inputs in a category, the category name and menu should be visible. This change breaks the desired behaviour and I think we should tweak it. Also, we may need a test for this?

Here is what it looked like before the change:
![image](https://user-images.githubusercontent.com/25869467/122056025-dfc56480-cdf1-11eb-91fa-54ec01f80375.png)

SaraBCL replied:
Good catch, that was bad ! Better now.

IvaKop commented on front/app/modules/commercial/insights/admin/containers/Insights/Edit/InputsTable/index.tsx:
```diff
@@ -105,20 +135,45 @@ const InputsTable = ({
+
+  const toggleInputSelected = (input: IInsightsInputData) => () => {
+    if (selectedRows.has(input.id)) {
+      const newSelection = new Set(selectedRows);
```
Do you need to define this twice in both if blocks or can it be put on top of the function?

SaraBCL replied:
Ah indeed :) Fixed.

IvaKop commented on front/app/modules/commercial/insights/admin/containers/Insights/Edit/InputsTable/index.tsx:
```diff
@@ -72,9 +95,16 @@ const InputsTable = ({
+  const selectedCategory = query.category;
+
+  // Reset selection on page change
+  useEffect(() => {
```
Clever! I also like the use of sets for this :)

IvaKop commented on front/app/modules/commercial/insights/services/batchAssignment.ts:
```diff
@@ -0,0 +1,49 @@
+    }
+  );
+  streams.fetchAllWith({
+    partialApiEndpoint: [`insights/views/${insightsViewId}/inputs`],
```
What does it mean partialApiEndpoint?

SaraBCL replied:
I think it means it'll also refetch urls containing that, for instance individual inputs.

IvaKop replied:
Ah, good to know. Will use that as well. Thanks!

IvaKop commented on front/app/modules/commercial/insights/services/batchAssignment.ts:
```diff
@@ -0,0 +1,49 @@
+    `${API_PATH}/${getInsightsBatchEndpoint(insightsViewId)}assign_categories`,
+    {
+      inputs,
+      categories,
```
This is not exactly a comment for this PR but when we delete categories, we don't send an array of strings but an object: `{. id: some id, type: category }`. I think we should be consistent and I like the array of id-s better, so I wonder if we should change the other endpoints?

IvaKop commented on front/app/modules/commercial/insights/admin/containers/Insights/Edit/InputsTable/InputsTableRow.tsx:
```diff
@@ -25,41 +25,44 @@ const CategoryList = styled.div`
+        <Checkbox checked={selected} onChange={changeSelected} />
       </td>
-      <td>
+      <td onClick={onPreview}>
```
I see why this was needed but can we fix it with stopPropagation() on the click event for the checkbox? It would be better if the entire row is clickable, I think.

SaraBCL replied:
I had to make a change to the checkbox component to achieve this (Since it was relying on the label click, stopping propagation wouldn't work). PR up soon

IvaKop commented on front/app/modules/commercial/insights/admin/containers/Insights/Edit/InputsTable/InputsTable.test.tsx:
```diff
@@ -150,6 +177,173 @@ describe('Insights Input Table', () => {
     );
   });
+
+  // Rows selection and actions
```
Let's add these types of comments to describe blocks? I think it will make it clearer with the actions, also with the pagination maybe?

SaraBCL replied:
I've tried to name the code blocks in the index file, hope that's what you meant. Super hard to do so, tough, with effects and callbacks that can't go after the null retrun for the hooks ordering, and the other functions tht rely on the data being there.

IvaKop replied:
Sorry, I should have been more clear. All I meant was to group related tests in 'describe' blocks nested inside the general describe block just to make it look tidier in the test file. Like:
```describe('Actions', ()=>{ ... tests ...})```
Moving things around in the index file would indeed be difficult and is also not super necessary, imo :)

IvaKop commented on front/app/modules/commercial/insights/admin/containers/Insights/Edit/InputsTable/InputsTable.test.tsx:
```diff
@@ -150,6 +177,173 @@ describe('Insights Input Table', () => {
+
+      render(<InputsTable />);
+      // Initially all rows are unchecked
+      expect(
```
Love it! You tested this in a really clever way 👍

SaraBCL replied:
:+1:

IvaKop commented on front/app/modules/commercial/insights/admin/containers/Insights/Edit/InputsTable/Actions.tsx:
```diff
@@ -0,0 +1,254 @@
+                    {otherCategories.map((category) => (
+                      <DropdownListItem
+                        key={category.id}
+                        onClick={toggleCategory(category.id)}
```
One little problem I noticed here is that because this onClick action is on the container and not the checkbox itself, if you click directly on it - it does not get selected.

SaraBCL replied:
Fixed, good catch. Working on the checkbox I noticed it should support mixed state, I'll log into tech debt to properly move the old mixed boxes to the compoent library

IvaKop reviewed (changes requested):
It looks really nice!!! 👍 

Two things I think we should fix before merging:
- When there are no inputs for a category, the category name and menu should still appear on top of the relevant table empty state
- Clicking directly on a checkbox in the bulk action menu does not select/deselect it.

SaraBCL committed:
Review fixes

SaraBCL committed:
Upgrades component library

SaraBCL committed:
Describe blocks for inputs table tests

SaraBCL committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into IN-401/bulk-table-actions

Commit:
Translations updated by CI (extract-intl)

SaraBCL committed:
Updates package lock to maybe fix build in CI

SaraBCL committed:
Removes unused import

SaraBCL committed:
Merge branch 'IN-401/bulk-table-actions' of github.com:CitizenLabDotCo/citizenlab into IN-401/bulk-table-actions

IvaKop commented:
This looks great! A couple of (non-blocking, I believe, comments):
- When I select a checkbox from the table, I briefly see the side-panel for some reason. I think it's easily fixed by stopping the propagation on checkbox click
- When I select items in the bulk action menu and I add them, when I open it again, they are still selected. Maybe the state should reset at that point?

SaraBCL committed:
Resets category selection on submit

SaraBCL commented:
Thanks :) 
- Did you install the new version of the component library ? I fixed it there, and I no longer experience the issue.
- Nice catch, fixed.

IvaKop commented:
Yes, it works perfect with the new library installed 👍 
Sorry for missing that :)
